### PR TITLE
UCT/CUDA/CUDA_IPC: cuda ipc uct skeleton

### PR DESCRIPTION
**Pull request by bureddy**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6655/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6660/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6662/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6665/ for details (Mellanox internal link).

File: src/uct/cuda/cuda_ipc/cuda_ipc_iface.c [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+    iface_attr->cap.get.max_bcopy       = 0;
+    iface_attr->cap.get.min_zcopy       = 0;
+    iface_attr->cap.get.max_zcopy       = 0;
+    iface_attr->cap.get.opt_zcopy_align = 1;
+    iface_attr->cap.get.align_mtu       = iface_attr->cap.get.opt_zcopy_align;
+    iface_attr->cap.get.max_iov         = 1;
+
+    iface_attr->cap.am.max_short        = 0;
+    iface_attr->cap.am.max_bcopy        = 0;
```
Comment (brminich) [inline src/uct/cuda/cuda_ipc/cuda_ipc_iface.c]: maybe no need to init AM related fields as you do not declare AM capability?

File: src/uct/cuda/cuda_ipc/cuda_ipc_iface.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+#define UCT_CUDA_IPC_TL_NAME    "cuda_ipc"
+#define UCT_CUDA_DEV_NAME   "cudaipc0"
+
+#define CUDA_FUNC(func)  ({                             \
+ucs_status_t _status = UCS_OK;                          \
+do {                                                    \
+    CUresult _result = (func);                          \
+    if (CUDA_SUCCESS != _result) {                      \
+        ucs_error("[%s:%d] cuda failed with %d \n",     \
```
Comment (brminich) [inline src/uct/cuda/cuda_ipc/cuda_ipc_iface.h]: AFAIR, ucs_error should print FILE and LINE itself

File: src/uct/cuda/cuda_ipc/cuda_ipc_md.c [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+                                         unsigned flags, uct_mem_h *memh_p)
+{
+    ucs_status_t rc;
+    uct_mem_h * mem_hndl = NULL;
+
+    mem_hndl = ucs_malloc(sizeof(void *), "cuda_ipc handle for test passing");
+    if (NULL == mem_hndl) {
+        ucs_error("Failed to allocate memory for gni_mem_handle_t");
+        rc = UCS_ERR_NO_MEMORY;
+        goto mem_err;
```
Comment (brminich) [inline src/uct/cuda/cuda_ipc/cuda_ipc_md.c]: maybe just return here?

File: src/uct/cuda/cuda_ipc/cuda_ipc_md.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * @brief cuda_ipc MD descriptor
+ */
+typedef struct uct_cuda_ipc_md {
+    struct uct_md super;   /**< Domain info */
+} uct_cuda_ipc_md_t;
+
+/**
+ * cuda ipc domain configuration.
```
Comment (brminich) [inline src/uct/cuda/cuda_ipc/cuda_ipc_md.h]: maybe add doxygen tag here as well for consistency?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6671/ for details (Mellanox internal link).

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.h [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+#include <ucs/type/class.h>
+
+
+typedef struct uct_cuda_ipc_ep_addr {
+    int                ep_id;
+} uct_cuda_ipc_ep_addr_t;
+
+typedef struct uct_cuda_ipc_ep {
+    uct_base_ep_t           super;
+    struct uct_cuda_ipc_ep *next;
```
Comment (yosefe) [inline src/uct/cuda/cuda_ipc/cuda_ipc_ep.h]: what is the purpose here? maybe can use ucs_list/ucs_queue here?

File: src/uct/cuda/cuda_ipc/cuda_ipc_iface.h [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+#define UCT_CUDA_IPC_IFACE_H
+
+#include <uct/base/uct_iface.h>
+#include <ucs/arch/cpu.h>
+#include <cuda_runtime.h>
+#include <cuda.h>
+
+
+#define UCT_CUDA_IPC_TL_NAME    "cuda_ipc"
+#define UCT_CUDA_DEV_NAME   "cudaipc0"
```
Comment (yosefe) [inline src/uct/cuda/cuda_ipc/cuda_ipc_iface.h]: pls indent on column

File: src/uct/cuda/cuda_ipc/cuda_ipc_iface.h [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+
+#define UCT_CUDA_FUNC(_func)  ({                        \
```
Comment (yosefe) [inline src/uct/cuda/cuda_ipc/cuda_ipc_iface.h]: copy/paste from codacopy, move to common header

File: src/uct/cuda/cuda_ipc/cuda_ipc_md.c [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+#include <ucs/sys/sys.h>
+#include <ucs/debug/memtrack.h>
+#include <ucs/type/class.h>
+#include <cuda_runtime.h>
+#include <cuda.h>
+
+
+static ucs_config_field_t uct_cuda_ipc_md_config_table[] = {
+    {"", "", NULL,
+        ucs_offsetof(uct_cuda_ipc_md_config_t, super), UCS_CONFIG_TYPE_TABLE(uct_md_config_table)},
```
Comment (yosefe) [inline src/uct/cuda/cuda_ipc/cuda_ipc_md.c]: pls fix indentation

File: src/uct/cuda/cuda_ipc/cuda_ipc_iface.h (hunk 7 above)
Comment (bureddy) [reply src/uct/cuda/cuda_ipc/cuda_ipc_iface.h]: this macro is not used in this PR.  removed it,  will move it to common header when it is used.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/6686/ for details (Mellanox internal link).

**Pull request merged**